### ISSUE TITLE
Custom config template name - my-serverless.yml

### ISSUE THREAD

**Issue by fivepapertigers** [serverless/serverless#4485]
# This is a Feature Proposal

## Description

For bug reports:
* What went wrong? *n/a*
* What did you expect should have happened? *n/a*
* What was the config you used? *n/a*
* What stacktrace or error message from your provider did you see? *n/a*

For feature proposals:
* What is the use case that should be solved. The more detail you describe this in the easier it is to understand for us.

On our projects, we frequently have multiple related serverless services (typically ones that touch shared stateful resources or align to similar business domains) in the same repo to maintain a consistent life-cycle across them. We separate them into multiple services for a number of reasons - sometimes they are logical separations (the services have fundamentally separate purposes), and other times it is a technical limitation (too many resources for a single Cloudformation stack). But what we end up with is two services that need access to the same codebase and often a shared set of configuration values. It would be nice for us to store multiple `serverless.yml` files side-by-side as such:
```
.
+-- serverless-api.yml
+-- serverless-workflow1.yml
+-- serverless-workflow2.yml
+-- app
|   +-- __init__.py
|   +-- py_module.py
```
* If there is additional config how would it look

I would imagine this could be easily implemented as a command-line option:
```bash
$ serverless deploy --configFile <path/to/config>
```

I would be more than happy to help make this feature a reality. A quick glance at the codebase makes it seem like the change would be relatively straight-forward.

Similar or dependent issues:
None that I could find (I dug for a while, I promise).

## Additional Data

* ***Serverless Framework Version you're using***:
* ***Operating System***:
* ***Stack Trace***:
* ***Provider Error messages***:

**Comment by horike37** [comment]
Thank you for the suggestion @fivepapertigers :+1: 
I like the idea. Selecting a serverless config file freely would make a serverless project based on the same codebase more extensible. 

I also agree with using a command-line option to pass a config file name.

**Comment by fivepapertigers** [comment]
@horike37 what's the process for making this happen? I'm very willing to write it myself, but want to get sign-off before starting that process and opening a PR.

**Comment by horike37** [comment]
Awesome @fivepapertigers 💯 It would be great if you will implement it!
I personally agree, but I also would like to discuss with someone else if there would not introduce any side effect by this feature.

Hi, @serverless/framework-maintainers, any concern on introducing this feature?

BTW, imo it would be good to the option name is `--config-file` instead of `--configFile` since many options of the framework follows hyphen scheme.

**Comment by horike37** [comment]
Hi, @fivepapertigers.
Your problem might have been resolved by the existing variables functionality.
If so, this feature doesn't need to be introduced.

You can do that also with e.g.:
```# serverless.yml
provider: ...
functions: ${file(${opt:config-file}):functions}
resources: ${file(${opt:config-file}):resources}
```
and then just invoke serverless as: `sls deploy --config-file <path/to/config>`.
The `opt` variable allows you to specify any custom arguments of the command line.
https://serverless.com/framework/docs/providers/aws/guide/variables#referencing-cli-options

Could you try it?

**Comment by HyperBrain** [comment]
Small addition 😸 : You should also define the service name in the referenced file and reference it with a variable out from there.

**Comment by fivepapertigers** [comment]
@horike37 That's a really cool way to accomplish it with the current feature set. It also allows commonalities to be abstracted into the parent serverless.yml file, which is really cool. I gave it a shot and seems like it's working well with one caveat: warnings appear if `${file(${opt:config-file}):plugins}` or a similar top-level attribute is not found or is empty in the child-config. This might happen if one child config requires the `plugins` attribute, but another config doesn't. Not a huge deal, since the warnings are non-blocking but still not ideal. I update here if I run into any other issues.

I would still argue that it makes a sense to have this as built-in functionality, though, as it's a common feature in many libraries (webpack, eslint, etc) - it would be especially familiar to JS developers. It's also seems like a relatively small, low-risk lift. And the proposed solution, while very creative, may feel a little indirect to someone who is just coming onto a project.

Thanks again, though, for giving me away to get off the ground with this. I'm excited to start implementing it.

**Comment by horike37** [comment]
Thank you for the response @fivepapertigers 👍 

>I would still argue that it makes a sense to have this as built-in functionality

OK, great!!
I agree with introducing this into the framework core 😄 

On the condition, the way I presented on above comment is that the settings need to be made common on parent serverless.yml file. Therefore, it would be very useful if you have multiple serverless projects which you can't commonalize in one repository.

Any opinions? @HyperBrain

**Comment by fivepapertigers** [comment]
I wanted to report back on this. I have been successfully able to support multiple stacks within the same directory using a custom option and a parent `serverless.yml` file as @horike37 has suggested. There are a couple of gotchas (e.g. all stacks must have the same top-level keys, you can't have one child stack with `resources` and another without), but that's mostly manageable. I'm happy to close this issue, as the workaround is fairly straight-forward and flexible.

**Comment by a1anw2** [comment]
@horike37 thank you for this tip.   Just ran into the CloudFormation limit and this really helped.

**Comment by et304383** [comment]
I'm shocked to find the config file name isn't an input value.  Wow.

**Comment by marcelpanse** [comment]
@horike37 solution doesn't work for plugins since the plugins are loaded before any templating happens in serverless. Currently, there doesn't seem a way to conditionally load some plugins. This simple config file proposal would fix it without changing the way serverless currently loads the plugins before any placeholders can be used.

**Comment by bawbgale** [comment]
The existing variables functionality also doesn't seem to work for the `provider` section for doing aws and azure deployments from the same repo.

If you try to include the entire provider section, it doesn't get evaluated (i.e. `srs print` just spits out the un-evaluated expression:
```
provider: ${file(serverless-${env:PROVIDER}.yml):provider}
```
Trying to parameterize each key doesn't work because it changes to order, which seems to cause the deploy to fail:
```
//serverless.yml
...
provider:
  name: ${file(serverless-${env:PROVIDER}.yml):provider.name}
  runtime: ${file(serverless-${env:PROVIDER}.yml):provider.runtime}
  stage: ${file(serverless-${env:PROVIDER}.yml):provider.stage}
...
```
```
> sls print
service: my-crossplatform-service
provider:
  stage: prod
  name: aws
  runtime: nodejs8.10
```

I ended up just deploying with a little bash script that copies the appropriate config file first:
```
#!/bin/bash
if [ "$1" != "" ]; then
  echo "copying serverless-$1.yml to serverless.yml and running serverless deploy"
  cp serverless-$1.yml serverless.yml && sls deploy
else
    echo "Please append provider, like 'deploy.sh aws' or 'deploy.sh azure'"
fi
```
But it would be so much easier just to specify the config file as a deploy option!

**Comment by fivepapertigers** [comment]
@bawbgale Agreed that the CLI flag would be easier and more robust. The fact that the workaround doesn't actually work around all use cases makes it frustrating. You end up refactoring your entire config only to find out that you can't accomplish an integral piece of your deployment needs.

**Comment by jgunn987** [comment]
Cannot believe this is not an option, very disappointing. How do i manage deployments to different environments? for now i will store each config outside the main repo, `cd` into the directory, copy in the config file i want to use into the root directory, run deploy, `rm` the config file i copied in, `git checkout` the repo to clean up. What a drag

**Comment by jgunn987** [comment]
I have a fork here https://github.com/jgunn987/serverless. this patches the getServelessConfig function to allowing you to specify a full path to a config file. Needs integration into the CLI though and more testing

**Comment by jgunn987** [comment]
Here is the PR https://github.com/serverless/serverless/pull/5589

**Comment by johndoe2361289638163** [comment]
Any update on this?

**Comment by bawbgale** [comment]
@johndoe2361289638163 Looks like updating docs is the main outstanding to do. @jgunn987 I've forked your fork and will take a stab at updating docs.

**Comment by dschep** [comment]
AFAICT this is a duplicate of #4473. gonna just leave it open tho since there's a pr against this issue

**Comment by AnthonyWC** [comment]
@bawbgale It doesn't work for provide: name section (e.g. aws/azure) but will work for other option (e.g. runtime).
